### PR TITLE
INFINITY-3007: Customized release indexing

### PR DESCRIPTION
**Pull request by nickbp**
- Release GAs at every 100 indexes by default, leaving space for hotfixes
- Release Betas sequentially (no index skipping)
- Support manual RELEASE_INDEX value for (GA) hotfixes

### CODE REVIEW

File: tools/universe/package_publisher.py [hunk 1]
```diff
@@ -11,27 +11,76 @@
...
+        if self._release_index >= 0:
+            if os.path.exists(os.path.join(repo_pkg_base, str(self._release_index))):
+                raise Exception('Requested index {} is already occupied in directory {}: {}'.format(
+                    self._release_index, repo_pkg_base, sorted(os.listdir(repo_pkg_base))))
+            this_index = self._release_index
+
+            # Search backwards from this_index to find a prior release to diff against:
+            for num in reversed(range(0, this_index)): # reversed([0, ..., this-1]) => [this-1, ..., 0]
+                if os.path.isdir(os.path.join(repo_pkg_base, str(num))):
+                    last_index = num
```
Comment (elezar) [inline tools/universe/package_publisher.py]: Is the diff not applicable in all cases and not only when we specify the release index?

File: tools/universe/package_publisher.py [hunk 2]
```diff
@@ -11,27 +11,76 @@
...
 
+    def _find_release_index(self, repo_pkg_base):
+        '''Returns the correct number/id for this release in the universe tree, and the prior release to diff against.
+
+        Returns a tuple containing two ints: [prior_index (or -1 if N/A), this_index]'''
+
+        # Find the index to be used for this release:
+        last_index = -1
+        if self._release_index >= 0:
+            if os.path.exists(os.path.join(repo_pkg_base, str(self._release_index))):
```
Comment (elezar) [inline tools/universe/package_publisher.py]: If we allow for indices to be specified explicitly, then even the automatic case could result in an overlap with an existing release. Should this check not be in the general path?

File: tools/universe/package_publisher.py [hunk 3]
```diff
@@ -11,27 +11,76 @@
...
+            this_index = self._release_index
+
+            # Search backwards from this_index to find a prior release to diff against:
+            for num in reversed(range(0, this_index)): # reversed([0, ..., this-1]) => [this-1, ..., 0]
+                if os.path.isdir(os.path.join(repo_pkg_base, str(num))):
+                    last_index = num
+                    break
+        else:
+            # Iterate over the directory contents and find the largest integer
+            for filename in os.listdir(repo_pkg_base):
```
Comment (elezar) [inline tools/universe/package_publisher.py]: Is there a way to restrict this to directories?

File: tools/universe/package_publisher.py (hunk 2 above)
Comment (nickbp) [reply tools/universe/package_publisher.py]: The automatic case is checking what's currently present in the repo to figure out the next number to use as 'latest'. Shouldn't have a conflict assuming things are added sequentially, i.e. unless someone has two parallel PRs out at the same time.

File: tools/universe/package_publisher.py (hunk 1 above)
Comment (nickbp) [reply tools/universe/package_publisher.py]: The diff is performed between last_index and this_index, and both the automatic and manual cases set these two values.

In the automatic case, last_index is found first (largest number in the package), and then this_index is calculated against that.

File: tools/universe/package_publisher.py (hunk 3 above)
Comment (nickbp) [reply tools/universe/package_publisher.py]: Added a directory check, mirroring the one a few lines up in the manual case. _Shouldn't_ be anything else here but can't hurt to be explicit..

**Pull request merged**